### ISSUE TITLE
Enable fatal warnings for Scala 3

### ISSUE THREAD

**Issue by armanbilge** [http4s/http4s#5184]
Came up in https://github.com/http4s/http4s/pull/5179#pullrequestreview-752091801. Scala 3 does have fatal warnings available, just not many warnings implemented to be fatal about :) things might improve in 3.1.0, not sure.

I'll look into this.

**Comment by rossabaker** [comment]
I don't think it hurts if it doesn't yet warn on much: we have extensive Scala 2 coverage for that.  It hurts if the warnings are spurious.  It helps if the warnings cover anything useful that Scala 2 doesn't.  I think the case elimination in #5179 was nice.

**Comment by bplommer** [comment]
The problem I've run into with this is that deprecation warnings work differently in Scala 2 and Scala 3. IIRC, Scala 2 doesn't warn about deprecated usages by code that is itself deprecated, but Scala 3 does. So you need `@nowarn` in scala 3, but not in scala 2.

**Comment by armanbilge** [comment]
Hmm. Except, I think Scala 3 does not have `@nowarn` yet.

**Comment by rossabaker** [comment]
I think it must: we're using `@nowarn` already.

If Scala 3 warns and Scala 2 doesn't, do we get into problems in Scala 2 of the `@nowarn` annotation not warning on anything and that being fatal?

**Comment by armanbilge** [comment]
Last I checked, `@nowarn` does nothing in Scala 3. Could be better now though.

> If Scala 3 warns and Scala 2 doesn't, do we get into problems in Scala 2 of the `@nowarn` annotation not warning on anything and that being fatal?

That's a good point.

**Comment by armanbilge** [comment]
Ok, I don't think this is going to work. For example:
```
[error] -- Error: /workspace/http4s/server/src/main/scala/org/http4s/server/middleware/CORS.scala:189:63 
[error] 189 |  def apply[F[_], G[_]](http: Http[F, G], config: CORSConfig = CORSConfig.default)(implicit
[error]     |                                                               ^^^^^^^^^^
[error]     |object CORSConfig in package org.http4s.server.middleware is deprecated since 0.21.27: Deficient. See https://github.com/http4s/http4s/security/advisories/GHSA-52cf-226f-rhr6.
[error] one error found
```

Which is annotated `@nowarn("cat=deprecation")`.
